extensions/amp-font/0.1/fontloader.js

```diff
@@ -0,0 +1,263 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @typedef {{
+ *  style: string,
+ *  variant: string,
+ *  weight: string,
+ *  size: string,
+ *  family: string
+ * }}
+ */
+var FontConfig;
+
+
+/** @private @const {Array.<string>} */
+const DEFAULT_FONTS_ = ['sans-serif', 'serif'];
+
+/** @private @const {string} */
+const TEST_STRING_ = 'MAxmTYklsjo190QW';
+
+/** @private @const {number} */
+const TOLERANCE_ = 2;
+
+
+import {removeElement} from '../../../src/dom';
+import {timer} from '../../../src/timer';
+import {vsync} from '../../../src/vsync';
+import * as style from '../../../src/style';
+
+
+export class FontLoader {
+
+  /**
+   * @param {!Document} doc
+   */
+  constructor(doc) {
+    /** @private @const {!Document} */
+    this.document_ = doc;
+    /** @private {?Element} */
+    this.container_ = null;
+    /** @private {?Array.<Element>} */
+    this.defaultFontElements_ = null;
+    /** @private {?Element} */
+    this.customFontElement_ = null;
+    /** @private {boolean} */
+    this.fontLoadResolved_ = false;
+    /** @private {boolean} */
+    this.fontLoadRejected_ = false;
+    /** @private {!FontConfig} */
+    this.fontConfig_ = null;
+  }
+
+
+  /**
+   * Triggers the font load. Returns promise that will complete when loading
+   * is considered to be complete.
+   * @param {!FontConfig} fontConfig Config that describes the font to be
+   *    loaded.
+   * @param {number} timeout number of milliseconds after which the font load
+   *    attempt would be stopped.
+   * @return {!Promise}
+   */
+  load(fontConfig, timeout) {
+    this.fontConfig_ = fontConfig;
+    return new Promise((resolve, reject) => {
+      timer.timeoutPromise(timeout, this.load_()).then(() => {
+        this.fontLoadResolved_ = true;
+        this.dispose_();
+        resolve();
+      }).catch(() => {
+        this.fontLoadRejected_ = true;
+        this.dispose_();
+        reject();
```

Comment:
When propagating errors don't forget to pass through the reason. It's in the `catch(reason)`. If you remove `new Promise` you can simply rethrow it as `throw reason`.